File: app/src/sharedTest/java/org/metabrainz/mobile/EntityTestUtils.java
package org.metabrainz.mobile;

import org.metabrainz.mobile.data.sources.api.entities.WikiSummary;
import org.metabrainz.mobile.data.sources.api.entities.mbentity.Artist;
import org.metabrainz.mobile.data.sources.api.entities.mbentity.Label;
import org.metabrainz.mobile.data.sources.api.entities.mbentity.Recording;
import org.metabrainz.mobile.data.sources.api.entities.mbentity.Release;
import org.metabrainz.mobile.data.sources.api.entities.mbentity.ReleaseGroup;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;

import static org.junit.Assert.fail;

public class EntityTestUtils {

    public static String loadResourceAsString(String resource) {
        StringBuilder builder = new StringBuilder();
        try (InputStream inputStream = ClassLoader.getSystemClassLoader().getResourceAsStream(resource);
             BufferedReader reader = new BufferedReader(new InputStreamReader(inputStream))) {
            String line;
            while ((line = reader.readLine()) != null)
                builder.append(line);
        } catch (IOException e) {
            e.printStackTrace();
            fail();
        }
        return builder.toString();
    }

    public static WikiSummary getTestArtistWiki() {
        WikiSummary summary = new WikiSummary();
        summary.setExtract("Edward Christopher Sheeran is an English singer, songwriter, record " +
                "producer, and actor. In early 2011, Sheeran independently released the extended " +
                "play, No. 5 Collaborations Project. After signing with Asylum Records, his debut " +
                "album, +, was released in September 2011 and topped the UK Albums Chart. It " +
                "contained his first hit single \"The A Team\". In 2012, Sheeran won the Brit Awards" +
                " for Best British Male Solo Artist and British Breakthrough Act.");
        return summary;
    }

    public static String getTestArtistMBID() {
        return "b8a7c51f-362c-4dcb-a259-bc6e0095f0a6";
    }

    public static Artist getTestArtist() {
        String MBID = getTestArtistMBID();
        Artist testArtist = new Artist();
        testArtist.setMbid(MBID);
        testArtist.setCountry("GB");
        testArtist.setDisambiguation("UK singer-songwriter");
        testArtist.setName("Ed Sheeran");
        testArtist.setSortName("Sheeran, Ed");
        testArtist.setGender("Male");
        return testArtist;
    }

    public static String getTestReleaseMBID() {
        return "94ad3a58-a1cc-46a3-acf4-9cb6c1d6f032";
    }

    public static Release getTestRelease() {
        String MBID = getTestReleaseMBID();
        Release testRelease = new Release();
        testRelease.setMbid(MBID);
        testRelease.setTitle("+");
        testRelease.setBarcode("5052498646524");
        testRelease.setStatus("Official");
        testRelease.setCountry("XE");
        testRelease.setDisambiguation("");
        testRelease.setDate("2011-09-09");
        return testRelease;
    }

    public static String getTestReleaseGroupMBID() {
        return "05ce100c-eddf-4967-8d7e-33fc0883fe39";
    }

    public static ReleaseGroup getTestReleaseGroup() {
        String MBID = getTestReleaseGroupMBID();
        ReleaseGroup testReleaseGroup = new ReleaseGroup();
        testReleaseGroup.setMbid(MBID);
        testReleaseGroup.setCount(1);
        testReleaseGroup.setTitle("+");
        testReleaseGroup.setPrimaryType("Album");
        testReleaseGroup.setDisambiguation("plus");
        return testReleaseGroup;
    }

    public static String getTestLabelMBID() {
        return "015a28ab-1eb8-45a0-b2c6-601e410548af";
    }

    public static Label getTestLabel() {
        String MBID = getTestLabelMBID();
        Label testLabel = new Label();
        testLabel.setMbid(MBID);
        testLabel.setName("Speed Records");
        testLabel.setCode("SPREC");
        testLabel.setCountry("IN");
        testLabel.setDisambiguation("India");
        testLabel.setType("Original Production");
        return testLabel;
    }

    public static String getTestRecordingMBID() {
        return "11f3c37d-839f-4f65-87c1-d55cb416d0c5";
    }

    public static Recording getTestRecording() {
        String MBID = getTestRecordingMBID();
        Recording testRecording = new Recording();
        testRecording.setMbid(MBID);
        testRecording.setLength(125400);
        testRecording.setTitle("Plus Plus");
        testRecording.setDisambiguation("no disambiguation");
        return testRecording;
    }
}

File: app/src/sharedTest/java/org/metabrainz/mobile/EntityTestUtils.kt
package org.metabrainz.mobile

import org.junit.Assert.fail
import org.metabrainz.mobile.data.sources.api.entities.ArtistCredit
import org.metabrainz.mobile.data.sources.api.entities.WikiSummary
import org.metabrainz.mobile.data.sources.api.entities.mbentity.*
import org.metabrainz.mobile.data.sources.api.entities.mbentity.Collection
import org.metabrainz.mobile.presentation.features.adapters.ResultItem
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.lang.StringBuilder
import java.util.ArrayList

object EntityTestUtils {

    fun loadResourceAsString(resource: String?): String {
        val builder = StringBuilder()
        try {
            ClassLoader.getSystemClassLoader().getResourceAsStream(resource).use { inputStream ->
                BufferedReader(
                    InputStreamReader(inputStream)
                ).use { reader ->
                    var line: String?
                    while (reader.readLine().also { line = it } != null) builder.append(line)
                }
            }
        } catch (e: IOException) {
            e.printStackTrace()
            fail()
        }
        return builder.toString()
    }

    val testArtistWiki: WikiSummary
        get() {
            val summary = WikiSummary()
            summary.extract =
                "Edward Christopher Sheeran is an English singer, songwriter, record " +
                        "producer, and actor. In early 2011, Sheeran independently released the extended " +
                        "play, No. 5 Collaborations Project. After signing with Asylum Records, his debut " +
                        "album, +, was released in September 2011 and topped the UK Albums Chart. It " +
                        "contained his first hit single \"The A Team\". In 2012, Sheeran won the Brit Awards" +
                        " for Best British Male Solo Artist and British Breakthrough Act."
            return summary
        }


    val testArtistMBID: String
        get() = "b8a7c51f-362c-4dcb-a259-bc6e0095f0a6"

    val testArtist: Artist
        get() {
            val MBID = testArtistMBID
            val testArtist = Artist()
            testArtist.mbid = MBID
            testArtist.type = "Person"
            testArtist.country = "GB"
            testArtist.disambiguation = "UK singer-songwriter"
            testArtist.name = "Ed Sheeran"
            testArtist.sortName = "Sheeran, Ed"
            testArtist.gender = "Male"
            val testReleases: MutableList<Release> = ArrayList()
            var testRelease = Release()
            testRelease.mbid = "8b6d9a7e-2f94-4cca-b2e3-d35116b1a49c"
            testRelease.date = "2005"
            testRelease.disambiguation = ""
            testRelease.title = "The Orange Room EP"
            testRelease.status = "Official"
            testRelease.country = "XW"
            testRelease.barcode = "8791423234"
            testReleases.add(testRelease)
            testRelease = Release()
            testRelease.mbid = "71c07e92-4b71-43d3-a041-5fd67d8343af"
            testRelease.date = "2006-03-22"
            testRelease.disambiguation = ""
            testRelease.title = "Ed Sheeran"
            testRelease.status = "Official"
            testRelease.country = "GB"
            testRelease.barcode = "12223213211"
            testReleases.add(testRelease)
            testRelease = Release()
            testRelease.mbid = "fe2c71cf-52e3-4abd-a0ed-45ab3770d965"
            testRelease.date = "2006-03-22"
            testRelease.disambiguation = ""
            testRelease.title = "Ed Sheeran (EP)"
            testRelease.status = "Official"
            testRelease.country = "GB"
            testRelease.barcode = "1221421"
            testReleases.add(testRelease)
            testRelease = Release()
            testRelease.mbid = "3c8473d3-202c-44c8-a37b-8e259fe12c39"
            testRelease.date = "2007-05-01"
            testRelease.disambiguation = ""
            testRelease.title = "Want Some?"
            testRelease.status = "Official"
            testRelease.country = "GB"
            testRelease.barcode = "8791423234"
            testReleases.add(testRelease)
            testArtist.setReleases(testReleases)
            return testArtist
        }

    val testReleaseMBID: String
        get() = "94ad3a58-a1cc-46a3-acf4-9cb6c1d6f032"

    val testRelease: Release
        get() {
            val MBID = testReleaseMBID
            val testRelease = Release()
            testRelease.mbid = MBID
            testRelease.title = "+"
            testRelease.barcode = "5052498646524"
            testRelease.status = "Official"
            testRelease.country = "XE"
            testRelease.disambiguation = ""
            testRelease.date = "2011-09-09"
            val testArtistCredits: ArrayList<ArtistCredit> = ArrayList<ArtistCredit>()
            val testArtistCredit = ArtistCredit()
            testArtistCredit.joinphrase = ""
            testArtistCredit.name = "Ed Sheeran"
            val testArtist = Artist()
            testArtist.mbid = "b8a7c51f-362c-4dcb-a259-bc6e0095f0a6"
            testArtistCredit.artist = testArtist
            testRelease.artistCredits = testArtistCredits
            return testRelease
        }

    val testReleaseGroupMBID: String
        get() = "05ce100c-eddf-4967-8d7e-33fc0883fe39"

    val testReleaseGroup: ReleaseGroup
        get() {
            val MBID = testReleaseGroupMBID
            val testReleaseGroup = ReleaseGroup()
            testReleaseGroup.mbid = MBID
            testReleaseGroup.count = 1
            testReleaseGroup.title = "+"
            testReleaseGroup.primaryType = "Album"
            testReleaseGroup.disambiguation = "plus"
            val testArtistCredits: List<ArtistCredit> = ArrayList<ArtistCredit>()
            val testArtistCredit = ArtistCredit()
            testArtistCredit.joinphrase = ""
            testArtistCredit.name = "Ed Sheeran"
            val testArtist = Artist()
            testArtist.mbid = "b8a7c51f-362c-4dcb-a259-bc6e0095f0a6"
            testArtistCredit.artist = testArtist
            testReleaseGroup.setArtistCredits(testArtistCredits)
            return testReleaseGroup
        }

    val testLabelMBID: String
        get() = "015a28ab-1eb8-45a0-b2c6-601e410548af"

    val testLabel: Label
        get() {
            val MBID = testLabelMBID
            val testLabel = Label()
            testLabel.mbid = MBID
            testLabel.name = "Speed Records"
            testLabel.code = "SPREC"
            testLabel.country = "IN"
            testLabel.disambiguation = "India"
            testLabel.type = "Original Production"
            return testLabel
        }

    val testRecordingMBID: String
        get() = "11f3c37d-839f-4f65-87c1-d55cb416d0c5"

    val testRecording: Recording
        get() {
            val MBID = testRecordingMBID
            val testRecording = Recording()
            testRecording.mbid = MBID
            testRecording.length = 125400
            testRecording.title = "Plus Plus"
            testRecording.disambiguation = "no disambiguation"
            val testArtistCredits: ArrayList<ArtistCredit> = ArrayList<ArtistCredit>()
            val testArtistCredit = ArtistCredit()
            testArtistCredit.joinphrase = ""
            testArtistCredit.name = "Like a Tim"
            val testArtist = Artist()
            testArtist.mbid = "90568a1e-ef8d-4d22-93fa-faadc819d576"
            testArtistCredit.artist = testArtist
            testRecording.artistCredits = testArtistCredits
            return testRecording
        }


    val testCollectionMBID: String
        get() = "a691377c-6949-44cb-8a10-9696178cca18"

    val testCollection: Collection
        get() {
            val MBID = testCollectionMBID
            val testCollection = Collection()
            testCollection.mbid = MBID
            testCollection.type = "Release"
            testCollection.name = "Rock and Roll"
            testCollection.count = 1
            testCollection.editor = "akshaaatt"
            testCollection.entityType = "release"
            return testCollection
        }

    val testCollectionDetails: ResultItem
        get() {
            TODO()
        }
}